Switch from libxml2 to lxml

nwellnhof opened the pull request:
Most of the transition is straight-forward, but some issues turned up.

- lxml doesn't seem to expose the prefixes of attributes, requiring an XPath evaluation as work-around.
- Serializing the internal subset is a bit hacky.
- lxml doesn't support attribute nodes, so we have to emulate them.
- lxml doesn't support attributes as XPath context nodes, so some use cases aren't supported. Using an `.` expression on an attribute works, though.

Changes to expected test results are mostly cosmetic.

- Whitespace before and after the document element is processed more faithfully.
- lxml removes some superfluous namespace prefixes.
- There's one superfluous namespace declaration which isn't removed anymore.

Fixes #10.

nwellnhof committed:
Switch from libxml2 to lxml

Most of the transition is straight-forward, but some issues turned up.

- lxml doesn't seem to expose the prefixes of attributes, requiring
  an XPath evaluation as work-around.
- Serializing the internal subset is a bit hacky.
- lxml doesn't support attribute nodes, so we have to emulate them.
- lxml doesn't support attributes as XPath context nodes, so some use
  cases aren't supported. Using an `.` expression on an attribute
  works, though.

Changes to expected test results are mostly cosmetic.

- Whitespace before and after the document element is processed more
  faithfully.
- lxml removes some superfluous namespace prefixes.
- There's one superfluous namespace declaration which isn't removed
  anymore.

Fixes #10.

shaunix commented:
This is really impressive. I was sure lxml didn't have the right hooks to make this possible. At a glance, the changes to the tests look superficial. For example, I don't think anybody cares about the newline before the timestamp comment. The namespace prefix changes might cause some churn in people's PO files. Not sure, I'd have to look more closely.

nwellnhof commented:
lxml has some idiosyncrasies but it allows to manipulate XML trees and has an XPath API which is basically all that's needed. It doesn't expose XML attributes as nodes of the tree, though, which makes things a bit complicated and leaves an open issue concerning attributes as XPath context nodes. But maybe it doesn't matter in practice.

I have only tested the PR against the test suite. It's a large commit and certainly needs more testing.

My main motivation is to deprecate and ultimately remove libxml2's original Python bindings. `itstool` seems to be the most prominent remaining user.

maurerle commented:
I did get this error trying to run itstool with pintail-itstool to build gnome-docs with this PR:

```
  File "env/bin/itstool", line 1548, in <module>
    doc.merge_translations(translations, opts.lang, strict=opts.strict)
  File "env/bin/itstool", line 872, in merge_translations
    self.generate_messages(comments=False)
  File "env/bin/itstool", line 1023, in generate_messages
    self.generate_message(self._doc.getroot(), None, comments=comments)
  File "env/bin/itstool", line 1129, in generate_message
    self.generate_message(child, msg, comments=comments, path=path)
  File "env/bin/itstool", line 1051, in generate_message
    msg.add_placeholder(node)
  File "env/bin/itstool", line 232, in add_placeholder
    holder = Placeholder(node)
  File "env/bin/itstool", line 194, in __init__
    self.name = ustr(xml_localname(node), 'utf-8')
  File "env/bin/itstool", line 382, in xml_localname
    return etree.QName(node.tag).localname
  File "src/lxml/etree.pyx", line 1872, in lxml.etree.QName.__init__
  File "src/lxml/apihelpers.pxi", line 1731, in lxml.etree._tagValidOrRaise
ValueError: Invalid tag name '<cyfunction Comment at 0x7f0cb2e45a40>'
```
I did not yet have time to look further into this issue.

nwellnhof committed:
Fix element check

nwellnhof commented:
@maurerle Can you try again with the commit above?

maurerle commented:
thanks @nwellnhof - this did not yet work:

I added 

```python
if not comments and isinstance(node, etree._Comment):
  return
```

in `    def generate_message(self, node, msg, comments=True, path=None):`

which seems to work.

Though I get this error twice while building all of Gnomes Docs:

```
Traceback (most recent call last):
  File "/env/gnome/bin/itstool", line 1555, in <module>
    doc.merge_translations(translations, opts.lang, strict=opts.strict)
  File "/env/gnome/bin/itstool", line 893, in merge_translations
    self.merge_translations(translations, language, node=child, strict=strict)
  File "/env/gnome/bin/itstool", line 893, in merge_translations
    self.merge_translations(translations, language, node=child, strict=strict)
  File "/env/gnome/bin/itstool", line 893, in merge_translations
    self.merge_translations(translations, language, node=child, strict=strict)
  File "/env/gnome/bin/itstool", line 895, in merge_translations
    newnode = self.get_translated(node, translations, strict=strict, lang=language)
  File "/env/gnome/bin/itstool", line 992, in get_translated
    ph_node = msg.get_placeholder(qname.localname).node
AttributeError: 'NoneType' object has no attribute 'node'
```

nwellnhof commented:
The old code was checking for elements:

        if node.type != 'element':
            return

> `AttributeError: 'NoneType' object has no attribute 'node'`

What are the steps to reproduce this error?

maurerle commented:
Well I am coming from here:

https://gitlab.gnome.org/Teams/Websites/help.gnome.org/-/issues/12#note_2438480

And these instructions somehow produce this error - though I am not yet deep into this.

The error comes from:

```
TRANS  hu /unstable/evolution/
Traceback (most recent call last):
  File "/home/maurer/.conda/envs/gnome/bin/itstool", line 1555, in <module>
    doc.merge_translations(translations, opts.lang, strict=opts.strict)
  File "/home/maurer/.conda/envs/gnome/bin/itstool", line 893, in merge_translations
    self.merge_translations(translations, language, node=child, strict=strict)
  File "/home/maurer/.conda/envs/gnome/bin/itstool", line 893, in merge_translations
    self.merge_translations(translations, language, node=child, strict=strict)
  File "/home/maurer/.conda/envs/gnome/bin/itstool", line 893, in merge_translations
    self.merge_translations(translations, language, node=child, strict=strict)
  File "/home/maurer/.conda/envs/gnome/bin/itstool", line 895, in merge_translations
    newnode = self.get_translated(node, translations, strict=strict, lang=language)
  File "/home/maurer/.conda/envs/gnome/bin/itstool", line 992, in get_translated
    ph_node = msg.get_placeholder(qname.localname).node
AttributeError: 'NoneType' object has no attribute 'node'
TRANS  oc /unstable/evolution/
Could not translate /unstable/evolution/ to hu
```

Might be that this did also not work with libxml2 - so I could also suspect this is an error in the hu translation file..?
Which is this one: https://gitlab.gnome.org/GNOME/evolution/-/blob/master/help/hu/hu.po?ref_type=heads

nwellnhof commented:
> Might be that this did also not work with libxml2 - so I could also suspect this is an error in the hu translation file..?

The error message seems to match the one from issue #1, so it's probably an incorrect placeholder.

nwellnhof commented:
For some reason, `etree.iselement` returns True for comments. I force-pushed a fix that should hopefully work.

maurerle commented:
Thanks! @nwellnhof this works just fine :)

diizzyy commented:
Hi,

I gave this a go on FreeBSD's ports tree and while it does seem to work looking at unit tests a lot of fallout says:

```
AttributeError: 'lxml.etree._Element' object has no attribute 'addContent'
```
Example: 
gucharmap 16.0.2

```
[ 63% 114/169] /usr/local/bin/itstool -m help/cs/gucharmap-cs.gmo --lang cs -o help/cs ../help/C/index.docbook ../help/C/legal.xml
FAILED: help/cs/index.docbook help/cs/legal.xml 
/usr/local/bin/itstool -m help/cs/gucharmap-cs.gmo --lang cs -o help/cs ../help/C/index.docbook ../help/C/legal.xml
Traceback (most recent call last):
  File "/usr/local/bin/itstool", line 1548, in <module>
    doc.merge_translations(translations, opts.lang, strict=opts.strict)
  File "/usr/local/bin/itstool", line 888, in merge_translations
    self.merge_credits(translations, language, node)
  File "/usr/local/bin/itstool", line 817, in merge_credits
    self._append_credits(self._itst_credits[0], node, trdata)
  File "/usr/local/bin/itstool", line 799, in _append_credits
    self._append_credits(newnode, child, trdata)
  File "/usr/local/bin/itstool", line 799, in _append_credits
    self._append_credits(newnode, child, trdata)
  File "/usr/local/bin/itstool", line 799, in _append_credits
    self._append_credits(newnode, child, trdata)
  File "/usr/local/bin/itstool", line 794, in _append_credits
    parent.addContent(val)
    ^^^^^^^^^^^^^^^^^
AttributeError: 'lxml.etree._Element' object has no attribute 'addContent'
```

This is using libxml2 2.14.3 with all later commits in the 2.14 branch patched in

nwellnhof committed:
Rewrite remaining call to `addContent`

nwellnhof commented:
That was a method call I missed to migrate. Can you try with the commit above?

diizzyy commented:
That helped, thanks!

Running it again show this error quite frequently in various projects:
```
ValueError: Invalid attribute name 'xml:lang'
```

Example:
drawing 1.0.2 (https://github.com/maoschanz/drawing/releases/tag/1.0.2)
```
[ 94% 34/34] /usr/local/bin/itstool -m help/ru/drawing-ru.gmo --lang ru -o help/ru ../help/C/fullscreen.page ../help/C/index.page ../help/C/legal.xml ../help/C/main_features.page ../help/C/preferences.page ../help/C/tools_classic.page ../help/C/tools_classic_colors.page ../help/C/tools_selection.page ../help/C/tools_transform.page ../help/C/whats_new.page ../help/C/zoom_preview.page
FAILED: help/ru/fullscreen.page help/ru/index.page help/ru/legal.xml help/ru/main_features.page help/ru/preferences.page help/ru/tools_classic.page help/ru/tools_classic_colors.page help/ru/tools_selection.page help/ru/tools_transform.page help/ru/whats_new.page help/ru/zoom_preview.page 
/usr/local/bin/itstool -m help/ru/drawing-ru.gmo --lang ru -o help/ru ../help/C/fullscreen.page ../help/C/index.page ../help/C/legal.xml ../help/C/main_features.page ../help/C/preferences.page ../help/C/tools_classic.page ../help/C/tools_classic_colors.page ../help/C/tools_selection.page ../help/C/tools_transform.page ../help/C/whats_new.page ../help/C/zoom_preview.page
Traceback (most recent call last):
  File "/usr/local/bin/itstool", line 1557, in <module>
    doc.merge_translations(translations, opts.lang, strict=opts.strict)
  File "/usr/local/bin/itstool", line 934, in merge_translations
    node.set(attr, language)
  File "src/lxml/etree.pyx", line 851, in lxml.etree._Element.set
  File "src/lxml/apihelpers.pxi", line 587, in lxml.etree._setAttributeValue
  File "src/lxml/apihelpers.pxi", line 1741, in lxml.etree._attributeValidOrRaise
ValueError: Invalid attribute name 'xml:lang'
```

nwellnhof committed:
Fix extraction of namespaced attributes from langPointer

nwellnhof commented:
See the newest commit. We should probably skip the `_itst_lang_attr` business if langPointer is `@xml:lang` which is the default, but I don't want to make any functional changes.

diizzyy commented:
Starting to look really good! Great work!

Another error that shows up in a handful of ports

```
AttributeError: 'NoneType' object has no attribute 'node'
```

Example: mate-terminal 1.28.1 (https://pub.mate-desktop.org/releases/1.28/)

```
if test -d "C"; then d="../"; else d="/wrkdirs/usr/ports/x11/mate-terminal/work/mate-terminal-1.28.1/help/"; fi; \
mo="ja/ja.mo"; \
if test -f "${mo}"; then mo="../${mo}"; else mo="/wrkdirs/usr/ports/x11/mate-terminal/work/mate-terminal-1.28.1/help/${mo}"; fi; \
(cd "ja/" && itstool -m "${mo}" ${d}/C/index.page ${d}/C/legal.xml ${d}/C/mate-terminal-contents.page ${d}/C/mate-terminal-encoding.page ${d}/C/mate-terminal-first-start.page ${d}/C/mate-terminal-introduction.page ${d}/C/mate-terminal-manage-profiles.page ${d}/C/mate-terminal-more-windows.page ${d}/C/mate-terminal-prefs-colors.page ${d}/C/mate-terminal-prefs-compatibility.page ${d}/C/mate-terminal-prefs-effects.page ${d}/C/mate-terminal-prefs-general.page ${d}/C/mate-terminal-prefs-scrolling.page ${d}/C/mate-terminal-prefs-title.page ${d}/C/mate-terminal-profiles.page ${d}/C/mate-terminal-reset.page ${d}/C/mate-terminal-resize.page ${d}/C/mate-terminal-shortcuts.page ${d}/C/mate-terminal-tabs.page ${d}/C/mate-terminal-title.page ${d}/C/mate-terminal-to-start.page ${d}/C/mate-terminal-windows.page) && \
touch "ja/ja.stamp"
if ! test -d "ka/"; then mkdir "ka/"; fi
if ! test -d "kab/"; then mkdir "kab/"; fi
Traceback (most recent call last):
  File "/usr/local/bin/itstool", line 1566, in <module>
    doc.merge_translations(translations, opts.lang, strict=opts.strict)
  File "/usr/local/bin/itstool", line 911, in merge_translations
    self.merge_translations(translations, language, node=child, strict=strict)
  File "/usr/local/bin/itstool", line 911, in merge_translations
    self.merge_translations(translations, language, node=child, strict=strict)
  File "/usr/local/bin/itstool", line 911, in merge_translations
    self.merge_translations(translations, language, node=child, strict=strict)
  File "/usr/local/bin/itstool", line 913, in merge_translations
    newnode = self.get_translated(node, translations, strict=strict, lang=language)
              ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/usr/local/bin/itstool", line 1010, in get_translated
    ph_node = msg.get_placeholder(qname.localname).node
              ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
AttributeError: 'NoneType' object has no attribute 'node'
```

nwellnhof commented:
AttributeError: 'NoneType' object has no attribute 'node'

I think these are caused by the missing placeholder issue #1, see my comment above. Can you check whether these errors occur with the old code as well?

nwellnhof committed:
Handle missing placeholders more gracefully

Print a warning to stderr instead of crashing.

This is not related to the lxml migration but to issue #1.

nwellnhof commented:
The newest commit is unrelated to the lxml migration and should fix the missing placeholder issue. Since this error seems to occur frequently, it might be useful remove that noise.

diizzyy commented:
itstool looks fine on my end after all fixes, many thanks for all your work you've put into this!

nwellnhof commented:
Thanks for testing the pull request!

satmandu commented:
Hello all, any chance of getting this merged soon?

maurerle commented:
@shaunix could you please take a look at this? :)

hongxu-jia commented:
Ping
